src/utils/markdown/components/rehype-transform.ts

```diff
@@ -102,31 +147,67 @@ export const rehypeTransformComponents: Plugin<
 				children: parent.children.slice(index + 1, indexEnd),
 			});
 
-			const replacementArray =
-				replacement instanceof Array ? replacement : [replacement];
-			// Replace child nodes (including comments) with the replacement component
-			parent.children.splice(
+			replacementMetas.push({
+				node,
 				index,
-				isRanged ? indexEnd - index + 1 : 1,
-				...(replacement ? (replacementArray as never) : []),
-			);
-
-			// Recursively transform the children
-			// This allows for nested components like ebook only content in tabs
-			replacementArray.forEach((replacement) => {
-				if (!isNodeParent(replacement)) return;
-				replacement?.children?.map((child) => {
-					const tree = { type: "root", children: [child] } as Root;
-					(
-						rehypeTransformComponents as (
-							props: RehypeComponentsProps,
-						) => (tree: Root, vfile: VFile) => void
-					)({ components })(tree, vfile);
-					return tree;
-				});
+				parent,
+				component,
+				componentName,
+				replacementProps,
+				isRanged,
+				indexEnd,
 			});
 
 			return;
 		});
+
+		await Promise.all([
+			...replacementMetas.map(async ({ componentName, component }) => {
+				const hasScript = await saveComponentScript(componentName, component);
+				if (!hasScript) return;
+				(vfile as MarkdownVFile).data.usedComponents[componentName] = true;
+			}),
+			...replacementMetas.map(
+				async ({
+					index,
+					parent,
+					component,
+					replacementProps,
+					isRanged,
+					indexEnd,
+				}) => {
+					const replacementFn = await getHastScriptCompFunction(
+						component.componentFSPath,
+					);
+
+					const replacement = replacementFn(replacementProps);
+
+					const replacementArray =
+						replacement instanceof Array ? replacement : [replacement];
+
+					// Recursively transform the children
+					// This allows for nested components like ebook only content in tabs
+					await Promise.all(
+						replacementArray.map(async (replacement) => {
+							if (!isNodeParent(replacement)) return;
+							await (
+								rehypeTransformComponents as (
+									props: RehypeComponentsProps,
+								) => (tree: Root, vfile: VFile) => void
```

Comment:
Should `rehypeTransformComponents` be cast like so? Consider one of the following 

If the final function returns a Promise, then it can be awaited
```diff
- (props: RehypeComponentsProps) => (tree: Root, vfile: VFile) => void;
+ (props: RehypeComponentsProps) => (tree: Root, vfile: VFile) => Promise<void>;
```
Otherwise, the await will have no effect, which will effectively return `Promise<void>`
```diff
-           await (
+           return (
								rehypeTransformComponents as (
									props: RehypeComponentsProps,
				 				) => (tree: Root, vfile: VFile) => void
			 				)({ components })(replacement as Root, vfile);
```